src/Kovan-DssSpell.t.sol

```diff
@@ -1278,71 +1653,382 @@ contract DssSpellTest is DSTest, DSMath {
+                } catch {}
+            }
+        }
 
-            hevm.warp(monday_1400_UTC - 3 days);                       // Friday,   14:00 UTC
-            assertEq(spell.nextCastTime(), monday_1400_UTC - 3 days);  // Able to cast
+        assertEq(vat.wards(address(end_old)), 0);
+        assertEq(cat.wards(address(end_old)), 0);
+        assertEq(vow.wards(address(end_old)), 0);
+        assertEq(pot.wards(address(end_old)), 0);
+        assertEq(spotter.wards(address(end_old)), 0);
+
+        assertEq(vat.wards(address(end)), 1);
+        assertEq(cat.wards(address(end)), 1);
+        assertEq(dog.wards(address(end)), 1);
+        assertEq(vow.wards(address(end)), 1);
+        assertEq(pot.wards(address(end)), 1);
+        assertEq(spotter.wards(address(end)), 1);
+
+        assertEq(end.wards(address(esmBug)), 1);
+        assertEq(end.wards(address(esmAttack)), 1);
+        assertEq(vat.wards(address(esmAttack)), 1);
+        assertEq(end_old.wards(address(esm_old)), 1);
+    }
 
-            hevm.warp(monday_2100_UTC);                                // Monday,   21:00 UTC
-            assertEq(spell.nextCastTime(), monday_1400_UTC + 1 days);  // Tuesday,  14:00 UTC
+    function testSpellIsCast_new_End_functionality() public {
+        vote(address(spell));
+        scheduleWaitAndCast(address(spell));
+        assertTrue(spell.done());
 
-            hevm.warp(monday_2100_UTC - 1 days);                       // Sunday,   21:00 UTC
-            assertEq(spell.nextCastTime(), monday_1400_UTC);           // Monday,   14:00 UTC
+        DSTokenAbstract ETH = DSTokenAbstract(addr.addr("ETH"));
+        GemJoinAbstract joinETHA = GemJoinAbstract(addr.addr("MCD_JOIN_ETH_A"));
+        FlipAbstract flipETHA = FlipAbstract(addr.addr("MCD_FLIP_ETH_A"));
+
+        DSTokenAbstract BAT = DSTokenAbstract(addr.addr("BAT"));
+        GemJoinAbstract joinBATA = GemJoinAbstract(addr.addr("MCD_JOIN_BAT_A"));
+        FlipAbstract flipBATA = FlipAbstract(addr.addr("MCD_FLIP_BAT_A"));
+
+        DSTokenAbstract LINK = DSTokenAbstract(addr.addr("LINK"));
+        GemJoinAbstract joinLINKA = GemJoinAbstract(addr.addr("MCD_JOIN_LINK_A"));
+        ClipAbstract clipLINKA = ClipAbstract(addr.addr("MCD_CLIP_LINK_A"));
+
+        uint256 ilkAmt = 1 * THOUSAND * WAD;
+
+        giveTokens(ETH, ilkAmt);
+        giveTokens(BAT, ilkAmt);
+        giveTokens(LINK, ilkAmt);
+
+        ETH.approve(address(joinETHA), ilkAmt);
+        joinETHA.join(address(this), ilkAmt);
+        BAT.approve(address(joinBATA), ilkAmt);
+        joinBATA.join(address(this), ilkAmt);
+        LINK.approve(address(joinLINKA), ilkAmt);
+        joinLINKA.join(address(this), ilkAmt);
+
+        (,uint256 rate, uint256 spot,,) = vat.ilks("ETH-A");
+        vat.frob("ETH-A", address(this), address(this), address(this), int256(ilkAmt), int256(mul(ilkAmt, spot) / rate));
+        uint256 totDaiGenerated = mul(ilkAmt, spot) / rate;
```

Comment:
I think this is giving you the total art generated instead of total dai. The art amount should always be less than the dai generated, so I think that's why it's not failing below.

Reply:
Ah yes I broke it meanwhile I was testing. I was multiplying by rate and dividing again to see if I had a rounding issue, then I forgot to add the division as well. Fixing it now.

Reply:
Done.

Reply:
I still don't think this is correct. It seems like this is a RAD amount and `end.pack()` wants a WAD. Maybe you could add some asserts `vat.dai(address(this), totDaiGenerated)` ?

Reply:
Well I guess the totDaiGenerated may be correct, but it should be divided by a RAY before being passed to the `end.pack()`.

Reply:
Yes you are right, it just needs to get divided by `RAY`. But I'm gonna use `vat.dai(address(this)) / RAY` directly. I could even use a random number lower than the actual balance. My point with this test is actually just check that all the functions work and not really testing any edge case. It's more a test to prove that all the contracts and authorities are set.

Reply:
I think there will be some rounding errors as well. It should maybe be `uint256 totDaiGenerated = (mul(ilkAmt, spot) / rate) * rate;`

Reply:
Fixed the test using `vat.dai(address(this))`. IMO this should be enough for this one.